keps/sig-architecture/4402-go-workspaces/README.md

```diff
@@ -0,0 +1,452 @@
+tests to tools and verifiers, and look closely at the results of the various
+code-generators.  Ideally there will be NO CHANGE in generated code.
+
+### Test Plan
+
+We will add some tests to untested tools, though complete functional coverage
+is difficult, and would be a prohibitively high bar (e.g. it's unreasonable to
+demand net-new tests for every codegen tool which exercises every path). That
+said, we will look carefully at how each tool works, to mitigate risks.  At the
+time of this writing, the author has spent upwards of 100 hours in the
+debugger, single-stepping through each tool.
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+We will add tests for critical subsystems and tools being changed (e.g. gengo's
+package loading).
+
+##### Integration tests
+
+N/A - see e2e
+
+##### e2e tests
+
+The best test for work like this is, more or less, "did it work?".
+
+If the codegen tools emit the exact same code (or changes are manually
+verifiable as correct) then "it works".  Part of this process must be to
+exercise it on out-of-tree consumers, which will be difficult to do until it is
+merged.  We expect some amount of ex post facto problem reports from
+downstreams.
+
+### Graduation Criteria
+
+N/A - once this merges it is live.
+
+### Upgrade / Downgrade Strategy
+
+N/A - not user-facing.
+
+### Version Skew Strategy
+
+N/A - not user-facing.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+N/A for clusters.  Once we merge it, it is active for developers in k/k.  Once
+we cut a release tag, it is active for consumers of k8s.io/code-generator.  If
+we should need to back it out, it will be a complex set of git reverts, across
+at least k/k and kube-openapi repos.
+
+As such, we must make sure that each PR works and is self-contained - no
+leaving the repo(s) in a broken state.
+
+###### Does enabling the feature change any default behavior?
+
+No.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+N/A.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+N/A.
+
+###### Are there any tests for feature enablement/disablement?
+
+N/A.
+
+### Rollout, Upgrade and Rollback Planning
+
+N/A.
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+N/A.
+
+###### What specific metrics should inform a rollback?
+
+If there is massive downstream breakage, we may need to revert. If we adopt Go
+1.22, merge this KEP, then decide to roll-back to Go 1.21 for some reason, we
+WILL need to revert this KEP.  We should have pretty high confidence in Go 1.22
+before we merge this.
```

Comment:
Probably would be good to ensure that at least the big consumers give it a try and report feedback. I'm happy to do that from the OpenShift side of things.

Reply:
I have a PR you can sync and try against...

https://github.com/kubernetes/kubernetes/pull/122624